### PR TITLE
doc : client team’s proposals on CLI and ressource/workload definition

### PR DESCRIPTION
**Pull request by NilsPonsard**
This is a duplicate of #5 and #7 but with a rewritten commit history, only one PR should be merged.

This pull request adds the documentation of the CLI we want to build plus the structure of resource and workload definition.

### CODE REVIEW

File: docs/proposals/cli.md [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+kudoctl delete instance id9898
+```
+  
+</details>
+
+---
+
+<details> <summary><h3>create &lt;kind&gt;</h3></summary>
+
+Create a resource definition. By default if a resource with the same name exists, the resource will be updated, add the `--no-update` flag if you don’t want this behavior.
```
Comment (sameo) [inline docs/proposals/cli.md:245]: If you want to support imperative mode and update, you should add an `update`command and avoid piggy backing the `create`one.

Comment (sameo) [inline docs/proposals/cli.md:245]: Will `create workload`create a worload resource but not a workload instance?

File: docs/proposals/cli.md [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+kudoctl apply -f workload.yml
+```
+
+</details>
+
+---
+  
+<details> <summary><h3>instantiate &lt;resource-id&gt;</h3></summary>
+
+Instantiate and start a resource
```
Comment (sameo) [inline docs/proposals/cli.md:311]: What does you mean by starting a resource? If a user is a resource, what does starting a user mean?

File: docs/proposals/cli.md (hunk 1 above)
Comment (sameo) [inline docs/proposals/cli.md:245]: The `create`semantic should be more clearly defined. You basically have 2 choices:

1. `create`means creating a resource and nothing else. If a resource with the same name exists, you get an error back. Instantiating a resource may not always make sense, so this is a slippery slope.
2. `create`means creating and instantiating (when it makes sense) a new resource. This is a more clearly defined semantic, an imperative one ("kudo, please create a new workload instance based on the workload definition that I'm giving you"

The second choice makes more sense and for certain resources (e.g. a workload) is equivalent to a declarative `apply`.

The bottom line is that you should go one way or the other (declarative or imperative) and my advice is that if you go declarative, you basically should only support apply and delete, not create.
If you go imperative, using nouns+verbs is a lot clearer, e.g. `workload create -f foo.yml` and then `workload start <workload id>`.

File: docs/proposals/cli.md [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+## Global flags
+
+| Name              | Shorthand | Values     | Default        | Description                                                                       |
+| ----------------- | --------- | ---------- | -------------- | --------------------------------------------------------------------------------- |
+| --Version         | -V        | bool       | false          | output the version of the client with the format : `v.major.minor.patch` (semver) |
+| --help            | -h        | bool       | false          | display the help text.                                                            |
+| --config          | -c        | path       | ~/.kudo/config | specify the path to the config file.                                              |
+| --host            |           | string     | `''`           | specify the ip of the control plane to connect to.                                |
+| --verbosity-level | -v        | int 0 to 3 | 2              | Set the verbosity level of the execution, see **Log format** section              |
```
[the remark this answers is not in the document]
Comment (NilsPonsard) [reply docs/proposals/cli.md:23]: Is it ok to just relpace the integer with a string like `--verbosity-level info` and `-v info` ?

File: docs/proposals/cli.md (hunk 1 above)
Comment (NilsPonsard) [reply docs/proposals/cli.md:245]: With Mathieu we decided to go declarative as you advised.

File: docs/proposals/cli.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sameo) [reply docs/proposals/cli.md:23]: That's equivalent to `log info`, so yes it's ok.

File: docs/proposals/cli.md [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+**Arguments :**  
+
+`id` : the id of the instance.
+
+**Flags :**  
+
+| Name      | Shorthand | Values                                       | Default            | Description                                                            |
+| --------- | --------- | -------------------------------------------- | ------------------ | ---------------------------------------------------------------------- |
+| ---format |           | `"json"`,`"human_readable"`,`"yaml"`,`"xml"` | `"human_readable"` | The output format of the resource definition,  yml is the same as yaml |
```
[the remark this answers is not in the document]
Comment (sameo) [reply docs/proposals/cli.md]: It's easier to *add* than to remove. Keep it simple for now and if there are some requests for `XML`(spoiler alert: There wont be) you can then add it.

**Review by sameo — COMMENTED** [review]
Looks mostly good to me, thanks.
Please address the remaining comment (re verbosity level) and squash your commits into one. Then we should be able to merge that PR.

**Review by sameo — APPROVED** [review]
